preprocessing/formparser.py

```diff
@@ -0,0 +1,847 @@
+                    # Read the service water demand for irrigation and livestock
+                    self.service_water_demand[name] = self.read_service_water(
+                        name, prefix[key], rainy_season
+                    )
+            else:
+                # Read the service water demand for service water
+                self.service_water_demand[key] = self.read_service_water(
+                    key, prefix[key]
+                )
+
+        return self.service_water_demand
+
+    def create_drinking_water_demand(self, drinking_prefix):
+        """
+        Create the drinking water demand for a given prefix.
+
+        Args:
+            drinking_prefix (str): The prefix for the drinking water demand.
+
+        Returns:
+            dict: A dictionary with the drinking water demand.
+        """
+        # Get the unit of measurement for the drinking water
+        unit_of_measurement = self.form[
+            f"{drinking_prefix}/drinking_express{self.suffix}"
+        ]
+        # Get the amount of water used
+        unit = float(self.form[f"{drinking_prefix}/drink_use{self.suffix}"])
+        # Get the time window for the drinking water usage
+        string_drink_window = self.form[f"{drinking_prefix}/drink_time{self.suffix}"]
+
+        buck_conversion = None
+        # Check if the unit of measurement is in buckets
+        if "buck" in unit_of_measurement:
+            buck_conversion = float(
+                self.form[f"{drinking_prefix}/drink_dim{self.suffix}"]
+            )
+        # Calculate the drinking water demand
+        consume = utils.convert_perliter(unit_of_measurement, unit, buck_conversion)
+        # Extract the time windows for the drinking water usage
+        drink_usage_time = utils.extract_time_windows(string_drink_window)
+
+        # Create the drinking water demand dictionary
+        self.drinking_water_demand = {"daily_demand": consume}
+        # Add the time window information to the dictionary
+        win, time_windows = utils.convert_usage_windows(drink_usage_time)
+
+        for key, item in win.items():
+            self.drinking_water_demand[f"water_{key}"] = item
+        return self.drinking_water_demand
+
+    def create_elec_appliance_demand(self, electric_prefix):
+        """
+        Create the electric appliance demand for a given prefix.
+
+        Args:
+            electric_prefix (str): The prefix for the electric appliance demand.
+
+        Returns:
+            dict: A dictionary with the electric appliance demand.
+        """
+        app_dict = {}
+
+        # Loop through every appliance
+        for key, data in self.form.items():
+            if electric_prefix in key and "_power" in key:
+                app_name = key.split(sep="/")[1].split(sep="_power")[0]
+
+                # Get the number of appliances
+                number = float(
+                    self.form[f"{electric_prefix}/{app_name}_number{self.suffix}"]
```

Comment:
is this key guaranteed to always be present? Or can the value associated not be castable to `float()` ? I am just looking ahead what type of deep errors could happen here. A large `try except KeyError` block which raises a comprehensive error message could make sense as we access a lot of keys inside the form dict.

Reply:
Actually I noticed further down that there is an elegant solution to this problem: using `self._read_form` instead of `self.form`